rpc, wallet: Add listaddresses RPC

lsilva01 opened the pull request:
This PR adds an address explorer RPC to Bitcoin Core, similar to "Address" tab of Electrum or Specter Desktop.

This allows the user to know any arbitrary range of external and internal addresses and can be used with legacy or descriptor wallet.

This same functionality can be used to build an "Address" tab in the GUI later.

For better visualization in manual tests, the reviewer can start the node with the keypool reduced to 2. (default is 1000) and create a new wallet.
```
./src/bitcoind -keypool=2
./src/bitcoin-cli -rpcwallet=<wallet_name> listaddresses
./src/bitcoin-cli -rpcwallet=<wallet_name> listaddresses "p2sh-segwit"
./test/functional/wallet_listaddresses.py --descriptors
./test/functional/wallet_listaddresses.py --legacy-wallet
```
<img width="1175" alt="image-2" src="https://user-images.githubusercontent.com/84432093/133831379-65c07e7d-d9f5-48dd-93ef-22838b91e9f2.png">

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -236,6 +236,128 @@ static void SetFeeEstimateMode(const CWallet& wallet, CCoinControl& cc, const Un
+                                {RPCResult::Type::STR, "hdkeypath", "The BIP 32 HD path of the address"},
+                                {RPCResult::Type::STR, "descriptor", "The address descriptor"},
+                                {RPCResult::Type::BOOL, "internal", "The address is internal (change) or external (receive)"},
+                                {RPCResult::Type::NUM, "amount", "Available amount at address"},
```
Addresses don't hold funds. They only receive, never send.

Instead, it should just tally the total received ever.

lsilva01 replied:
The code to count the total received is already available in the function. It can be done.

But address explorers in most software wallets, such as Electrum and Specter Desktop, only show the funds available to be spent at each address. I think the idea is to see how the UTXO values are distributed across the addresses (if there is any significant gap, for example).

achow101 commented:
Approach NACK

I started reviewing, but I have some fundamental issues with the approach taken. I like the general idea, but this specific implementation has issues.

First of all, why does this even need to take parameters? I don't see why a "starting index" or "ending index" need to be provided. We shouldn't need to derive addresses on the fly because they should already be in the wallet in some way. This really should use the address book, the keypool, and key metadata rather than re-deriving. Index just doesn't really make sense for what this RPC purports to do. This also ends up making assumptions about what addresses are receiving and change, depending on how they are derived. But that is not reliable, especially for wallets prior to the HD chain split. In fact, using this with such a wallet will result in an assertion failure, and that is just not acceptable. The correct thing is to use the wallet's built in change determining function(s) so that those that are listed as change actually correspond with the things that the wallet will see as change.

Likewise, why is this limited to just the default address type? A wallet will typically have addresses of address types that are not the default. These should be output as well.

Furthermore, this RPC appears to only work on HD wallets, which severely limits is usability. Many people do use non-HD wallets still, and I would expect something that just lists out the wallet's addresses is not limited to just HD wallets.

lsilva01 commented:
@achow101 thanks for detailed review.

My original implementation used the keypool (for legacy wallets), m_wallet_descriptor (for descriptor wallets) and key metadata.
Originally, the `ScriptPubKeyMan::ListAddresses()` methods also didn't call `TopUp()`. 

But the problem is the following case: if the user wants to know some address whose index is greater than the keypool total (such as 2395th), an error would be generated. Deriving addresses solved this issue. The listaddresses RPC would work regardless of keypool.

Since the problem was in the approach, not in concept, I changed the code of this PR to the original version, that uses the keypool, m_wallet_descriptor and key metadata.
So the way I've found to resolve the keypool limitation is the following code, but I'm not sure this is a good user experience.

```c++
if ((size_t) end_index >= setKeyPool->size()) {
    auto s_end_index = strprintf("%u", end_index);
    auto s_set_keypool_size = strprintf("%u", setKeyPool->size() - 1);
    auto error_message = "Error: end_index (" + s_end_index + ") is greater than last keypool item (" + s_set_keypool_size + ").";
    error = _(error_message.c_str());
    return false;
}
```

I didn't understand the reason to remove `start_index` and `end_index`. They work more like pagination. Without them, the command would return 2000 items (the default keypool size * 2). 
Removing them, what would be the other way to have a good visualization of the result?

I also added address type as parameter.

achow101 commented:
> But the problem is the following case: if the user wants to know some address whose index is greater than the keypool total (such as 2395th), an error would be generated. Deriving addresses solved this issue. The listaddresses RPC would work regardless of keypool size.

I don't think that is a situation we should support. It would mean that users could get addresses that are not currently being watched by the wallet. I think that there would be the expectation that the wallet is watching such an address and so any funds sent to them would appear in the wallet. But that would not be the case.

> I didn't understand the reason to remove `start_index` and `end_index`. They work more like pagination. Without them, the command would return 2000 items (the default keypool size * 2).

I think that it is completely reasonable to return such a large result. There could be options for verbosity, e.g. a non-verbose output is just the addresses, then verbose is with all of the additional information that this PR currently includes.

lsilva01 commented:
I removed the indexes and changed the code to use address book, the keypool, and key metadata.
The results of this RPC are:

. Legacy wallet:  The addresses from `address book` (those returned by `getnewaddress`) are not ordered, but the those from  keypool are in the correct BIP32 derivation order.  This is a limitation on legacy wallet result, since keypool erases the key in `getnewaddress`. I'm not sure there is a simple way to order them.

. Descriptor wallet: All the addresses are in the correct BIP32 derivation order, as it is possible to iterate over the `m_wallet_descriptor.cache` using `m_wallet_descriptor.range_start` and `m_wallet_descriptor.range_end`. 

Updated test information in PR description.

lsilva01 committed:
Add listaddresses RPC

lsilva01 committed:
Add functional test for listaddresses RPC

lsilva01 committed:
Change getaddressinfo() to use AddDestinationInfo() function

getaddressinfo() and listddresses() have fields in common.
AddDestinationInfo() was created to reuse the same code for both.
In this commit, getaddressinfo() is changed to call this function.

lsilva01 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -236,6 +236,103 @@ static void SetFeeEstimateMode(const CWallet& wallet, CCoinControl& cc, const Un
+
+    UniValue result(UniValue::VOBJ);
+
+    for (bool internal: { false, true }) {
```
Done in 4c2625a .

lsilva01 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1613,6 +1613,48 @@ std::set<CKeyID> LegacyScriptPubKeyMan::GetKeys() const
+
+    addresses.reserve(m_wallet_descriptor.range_end);
+
+    for (int32_t i = m_wallet_descriptor.range_start; i < m_wallet_descriptor.range_end; ++i) {
```
Done in 4c2625a.

`m_map_script_pub_keys` is sorted by descriptor range index before iteration.

lsilva01 commented on src/wallet/wallet.cpp:
```diff
@@ -2137,6 +2137,94 @@ bool CWallet::TopUpKeyPool(unsigned int kpSize)
+    }
+
+    for(auto destination : destinations) {
+        auto meta = spk_man->GetMetadata(destination);
```
Done in fcade9b.

Created a new function that both `getaddressinfo` and `listaddresses` call.
This new function keeps the order of `getaddressinfo` fields.

MarcoFalke commented:
Closing for now. Let us know when you are working on this again and if it should be reopened.
